### PR TITLE
Ed25519 to X25519

### PR DESCRIPTION
**Pull request by bgrieder**
Add support from the conversion of an Ed25519 Keypair to an X25519 Keypair.

See the test in ECIES Salsa Sealed Box for usage

### CODE REVIEW

File: src/ecies/ecies_salsa_sealed_box.rs [hunk 1]
```diff
@@ -133,6 +133,33 @@ mod tests {
...
+            plaintext.len() + EciesSalsaSealBox::ENCRYPTION_OVERHEAD
+        );
+
+        // generate the corresponding X25519 private key from the ED25519 private key
+        let x25519_sk = X25519PrivateKey::from_ed25519_private_key(&ed25519_sk);
+
+        // decrypt
+        let plaintext_ = EciesSalsaSealBox::decrypt(&x25519_sk, &ciphertext, None).unwrap();
+        // assert
+        assert_eq!(plaintext, &plaintext_[..]);
```
Comment (Manuthor) [inline src/ecies/ecies_salsa_sealed_box.rs:160]:
```suggestion
        let plaintext = EciesSalsaSealBox::decrypt(&x25519_sk, &ciphertext, None).unwrap();
        // assert
        assert_eq!(plaintext, &plaintext[..]);
```

**Review by Manuthor — COMMENTED** [review]
thanks for this conversion

File: src/ecies/ecies_salsa_sealed_box.rs (hunk 1 above)
Comment (bgrieder) [reply src/ecies/ecies_salsa_sealed_box.rs:160]: plaintext_ (with underscore) is the one recovered. plaintext (without underscore is the original)

File: src/asymmetric_crypto/curve_25519/x25519/mod.rs [hunk 2]
```diff
@@ -1,7 +1,29 @@
 mod public_key;
 
 pub use public_key::{X25519PublicKey, X25519_PUBLIC_KEY_LENGTH};
+use sha2::{Digest, Sha512};
 
-use super::private_key::Curve25519PrivateKey;
+use crate::{Ed25519PrivateKey, RandomFixedSizeCBytes, CURVE_25519_PRIVATE_KEY_LENGTH};
 
-pub type X25519PrivateKey = Curve25519PrivateKey;
+use super::private_key::Curve25519Secret;
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curve_25519/x25519/mod.rs:8]: yes

File: src/ecies/ecies_salsa_sealed_box.rs (hunk 1 above)
Comment (Manuthor) [reply src/ecies/ecies_salsa_sealed_box.rs:160]: oops.. ok

File: src/asymmetric_crypto/curve_25519/x25519/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curve_25519/x25519/mod.rs]: Done

**Review by Manuthor — DISMISSED** [review]

File: src/ecies/ecies_salsa_sealed_box.rs [hunk 3]
```diff
@@ -104,14 +104,14 @@ impl Ecies<X25519PrivateKey, X25519PublicKey> for EciesSalsaSealBox {
...
-
     use crate::{
-        asymmetric_crypto::{X25519PrivateKey, X25519PublicKey},
+        asymmetric_crypto::{Ed25519PrivateKey, X25519PrivateKey, X25519PublicKey},
         ecies::ecies_salsa_sealed_box::EciesSalsaSealBox,
-        CsRng, Ecies, FixedSizeCBytes, RandomFixedSizeCBytes,
+        CsRng, Ecies, Ed25519PublicKey, FixedSizeCBytes, RandomFixedSizeCBytes,
     };
 
+    use rand_chacha::rand_core::SeedableRng;
```
Comment (tbrezot) [inline src/ecies/ecies_salsa_sealed_box.rs:113]: Use the reexport instead?
```suggestion
    use crate::reexport::rand_core::SeedableRng;
```

File: src/asymmetric_crypto/curve_25519/private_key.rs [hunk 4]
```diff
@@ -15,11 +15,11 @@ pub const CURVE_25519_PRIVATE_KEY_LENGTH: usize = 32;
 /// but rather re-used as a base type for other final types on the curve
 /// such as `X25519PrivateKey`.
 #[derive(Hash, Clone, Debug, PartialEq, Eq, Zeroize, ZeroizeOnDrop)]
-pub struct Curve25519PrivateKey(pub(crate) [u8; CURVE_25519_PRIVATE_KEY_LENGTH]);
+pub struct Curve25519Secret(pub(crate) [u8; CURVE_25519_PRIVATE_KEY_LENGTH]);
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curve_25519/private_key.rs:18]: The reason is that the value held is not the private key value: it is the value the private key is derived from:
- x25519: there is some clamping to the curve done first
- Ed25519: there is a hash followed by clamping

Renaming the length and filename: yes. Done

File: src/asymmetric_crypto/curve_25519/x25519/private_key.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    asymmetric_crypto::curve_25519::private_key::Curve25519Secret, Ed25519PrivateKey,
+    RandomFixedSizeCBytes, CURVE_25519_PRIVATE_KEY_LENGTH,
+};
+
+pub type X25519PrivateKey = Curve25519Secret;
+
+impl X25519PrivateKey {
+    /// Convert the ED25519 private key to an X25519 private key
+    ///
+    /// This method is useful when an Ed25519 private key has been converted to an X25519 public key.
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curve_25519/x25519/private_key.rs:13]: Yes. Thanks

File: src/asymmetric_crypto/curve_25519/x25519/private_key.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    /// It will generate the private key that corresponds to the generated X25519 public key
+    /// from the original private key.
+    ///
+    /// See [`X25519PublicKey::from_ed25519_public_key`] for more details.
+    pub fn from_ed25519_private_key(sk: &Ed25519PrivateKey) -> Self {
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/asymmetric_crypto/curve_25519/x25519/private_key.rs:18]: Done.

File: src/ecies/ecies_salsa_sealed_box.rs (hunk 3 above)
Comment (bgrieder) [reply src/ecies/ecies_salsa_sealed_box.rs:113]: Yes. Fixed the doctest too

File: Cargo.toml [hunk 7]
```diff
@@ -47,16 +47,29 @@ chacha20 = { version = "0.9", optional = true }
...
 leb128 = { version = "0.2", optional = true }
-pkcs8 = { version = "0.10", features = ["std", "alloc", "pem"], optional = true }
+pkcs8 = { version = "0.10", features = [
+  "std",
+  "alloc",
+  "pem",
+], optional = true }
 rand_chacha = "0.3"
 rand_core = { version = "0.6.4", features = ["getrandom"] }
+sha2 = "0.10"
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply Cargo.toml:63]: Yes. linking it to feature curve25519

**Review by tbrezot — DISMISSED** [review]

**Pull request merged**